contracts/ReverseFullCheckpoint.sol

```diff
@@ -0,0 +1,427 @@
+    event SubnetBlockFinalized(bytes32 blockHash, int256 number);
+
+    function init(
+        bytes memory v2esbnHeader,
+        uint64 initEpoch,
+        int256 v2esbn
+    ) public {
+        require(INIT_STATUS == 0, "Already init");
+
+        bytes32 v2esbnHeaderHash = keccak256(v2esbnHeader);
+
+        address[] memory next = HeaderReader.getEpoch(v2esbnHeader);
+
+        require(next.length > 0, "No Epoch Validator Empty");
+
+        HeaderReader.ValidationParams memory v2esbnBlock = HeaderReader
+            .getValidationParams(v2esbnHeader);
+
+        require(v2esbnBlock.number == v2esbn, "Invalid Init Block");
+
+        headerTree[v2esbnHeaderHash] = Header({
+            receiptRoot: v2esbnBlock.receiptRoot,
+            stateRoot: v2esbnBlock.stateRoot,
+            transactionsRoot: v2esbnBlock.transactionsRoot,
+            parentHash: v2esbnBlock.parentHash,
+            mix: (uint256(v2esbnBlock.number) << 128) |
+                (uint256(v2esbnBlock.roundNumber) << 64) |
+                uint256(block.number)
+        });
+        validators[v2esbn] = Validators({
+            set: next,
+            threshold: int256((next.length * 2 * 100) / 3)
+        });
+        currentValidators = validators[v2esbn];
+        setLookup(next);
+        latestBlock = v2esbnHeaderHash;
+        latestFinalizedBlock = v2esbnHeaderHash;
+
+        committedBlocks[v2esbn] = v2esbnHeaderHash;
+        INIT_EPOCH = initEpoch;
+        INIT_STATUS = 1;
+        INIT_V2ESBN = uint64(int64(v2esbn));
+    }
+
+    /*
+     * @description core function in the contract, it can be summarized into three steps:
+     * 1. Verify subnet header meta information
+     * 2. Verify subnet header certificates
+     * 3. (Conditional) Update Committed Status for ancestor blocks
+     * @param list of rlp-encoded block headers.
+     */
+    function receiveHeader(bytes[] calldata headers) public {
+        for (uint256 x = 0; x < headers.length; x++) {
+            HeaderReader.ValidationParams memory validationParams = HeaderReader
+                .getValidationParams(headers[x]);
+
+            address[] memory next = HeaderReader.getEpoch(headers[x]);
+
+            // Verify subnet header meta information
+            require(validationParams.number > 0, "Repeated Genesis");
+            require(
+                validationParams.number >
+                    int256(
+                        uint256(
+                            uint64(headerTree[latestFinalizedBlock].mix >> 128)
+                        )
+                    ),
+                "Old Block"
+            );
+            require(
+                headerTree[validationParams.parentHash].mix != 0,
+                "Parent Missing"
+            );
+            require(
+                int256(
+                    uint256(
+                        uint64(
+                            headerTree[validationParams.parentHash].mix >> 128
```

Comment:
I have a question. there are many `headerTree[validationParams.parentHash]`. If we define a variable to store it, will it reduce gas? Like this
```
x = headerTree[validationParams.parentHash]
require(x.mix!=0)
require(...)
```